crypto/verification.go

```diff
@@ -0,0 +1,589 @@
+		mach.CancelSASVerification(userID, device.DeviceID, content.TransactionID, "Mismatched PK MACs", event.VerificationCancelKeyMismatch)
+		return
+	}
+
+	// we can finally trust this device
+	device.Trust = TrustStateVerified
+	mach.CryptoStore.PutDevice(device.UserID, device)
+
+	mach.Log.Debug("Device %v of user %v verified successfully!", device.DeviceID, device.UserID)
+}
+
+// handleVerificationCancel handles an incoming m.key.verification.cancel message.
+// It cancels the verification process for the given reason.
+func (mach *OlmMachine) handleVerificationCancel(userID id.UserID, content *event.VerificationCancelEventContent) {
+	// make sure to not reply with a cancel to not cause a loop of cancel messages
+	// this verification will get canceled even if the senders do not match
+	mach.keyVerificationTransactionState.Delete(userID.String() + ":" + content.TransactionID)
+	mach.Log.Warn("SAS verification %v was canceled by %v with reason: %v (%v)",
+		content.TransactionID, userID, content.Reason, content.Code)
+}
+
+// handleVerificationRequest handles an incoming m.key.verification.request message.
+func (mach *OlmMachine) handleVerificationRequest(userID id.UserID, content *event.VerificationRequestEventContent) {
+	mach.Log.Debug("Received verification request from %v", content.FromDevice)
+	otherDevice, err := mach.GetOrFetchDevice(userID, content.FromDevice)
+	if err != nil {
+		mach.Log.Error("Could not find device %v of user %v", content.FromDevice, userID)
+		return
+	}
+	// check if SAS is in supported methods
+	supportsSAS := false
+	for _, method := range content.Methods {
+		if method == event.VerificationMethodSAS {
+			supportsSAS = true
+			break
+		}
+	}
+	if !supportsSAS {
+		mach.Log.Warn("Canceling verification transaction %v as SAS is not supported", content.TransactionID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Only SAS method is supported", event.VerificationCancelUnknownMethod)
+		return
+	}
+	if mach.AcceptVerificationFrom(otherDevice) {
+		mach.Log.Debug("Accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		if err := mach.NewSASVerificationWith(otherDevice, content.TransactionID, 10*time.Minute, true); err != nil {
+			mach.Log.Error("Error accepting SAS verification request: %v", err)
+		}
+	} else {
+		mach.Log.Debug("Not accepting SAS verification %v from %v of user %v", content.TransactionID, otherDevice.DeviceID, otherDevice.UserID)
+		mach.CancelSASVerification(otherDevice.UserID, otherDevice.DeviceID, content.TransactionID, "Not accepted by user", event.VerificationCancelByUser)
+	}
+}
+
+// NewSASVerificationWith starts the SAS verification process with another device.
+// If the transaction ID is empty, a new one is generated.
+func (mach *OlmMachine) NewSASVerificationWith(device *DeviceIdentity, transactionID string, timeout time.Duration, useEmojiSAS bool) error {
```

Comment:
It would probably be nicer if the three last arguments were in an optional struct (see how [`SendMessageEvent`](https://github.com/tulir/mautrix-go/blob/master/client.go#L587) has an "optional" `extra` parameter). Also, timeout should have some default value and emojis should be on by default (maybe it should be disableEmojis instead of useEmojis)